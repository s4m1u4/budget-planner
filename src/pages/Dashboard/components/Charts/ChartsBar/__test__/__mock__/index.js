export { TITLE, DATA } from "./mock";
