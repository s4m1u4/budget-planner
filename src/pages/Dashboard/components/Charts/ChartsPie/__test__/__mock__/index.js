export { TITLE, DATA_INCOME } from "./mock";
