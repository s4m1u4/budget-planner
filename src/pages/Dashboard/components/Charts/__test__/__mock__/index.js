export { CATEGORIES, RECORDS } from "./mock";
