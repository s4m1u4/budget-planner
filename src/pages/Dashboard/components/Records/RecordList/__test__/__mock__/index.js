export { LAST_RECORDS } from "./mock";
