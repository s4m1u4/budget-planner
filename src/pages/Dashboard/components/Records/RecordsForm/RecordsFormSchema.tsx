import * as yup from "yup";

export const RecordsFormSchema = yup.object({
  amount: yup
    .number()
    .min(1)
    .positive("Amount should be a positive")
    .required("Amount is required"),
  type: yup.string().required("Type is required"),
  category: yup.string().required("Category is required"),
});
