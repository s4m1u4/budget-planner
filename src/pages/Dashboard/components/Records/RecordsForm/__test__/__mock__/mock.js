export const CATEGORIES = [
  {
    id: "084a3558-6b14-4e38-9d24-47e912021384",
    title: "Work",
    description: "Work",
    isDefault: false,
  },
  {
    id: "0ce314b9-a547-4f57-9d64-e466777ba8e9",
    title: "Travel",
    description: "awdawda",
    isDefault: false,
  },
  {
    id: "ea298d16-8bdb-4c5b-8198-6b2e5ce2f9a2",
    title: "Education",
    description: "Education it's so important for you!",
    isDefault: false,
  },
];
