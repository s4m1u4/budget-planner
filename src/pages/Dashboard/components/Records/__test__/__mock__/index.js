export { CATEGORIES, LAST_RECORDS } from "./mock";
