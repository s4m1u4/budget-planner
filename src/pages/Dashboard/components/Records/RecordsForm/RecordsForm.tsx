import React, { FC } from "react";
import { ModalComponent } from "components/shared";
import { Button, MenuItem, TextField } from "@mui/material";
import { useFormik } from "formik";
import { RecordsFormSchema } from "./RecordsFormSchema";
import { IHistoryData } from "../../../types";
import { ICategory } from "types";

import { ButtonGroup, Form, Title } from "./RecordsForm.styles";

interface RecordsFormProps {
  open: boolean;
  onSubmit: (values: IHistoryData) => void;
  categories: ICategory[];
  handleClose: () => void;
  setNewHistory: (historyData: IHistoryData) => void;
}

export const RecordsForm: FC<RecordsFormProps> = ({
  open,
  onSubmit,
  categories,
  handleClose,
  setNewHistory,
}) => {
  const handleSubmit = async (values: IHistoryData) => {
    await setNewHistory(values);
    await onSubmit(values);
  };

  const formik = useFormik({
    initialValues: {
      amount: "",
      type: "",
      category: "",
    },
    validationSchema: RecordsFormSchema,
    onSubmit: async (values: IHistoryData) => {
      await handleSubmit(values);
      formik.resetForm();
      handleClose();
    },
  });

  return (
    <ModalComponent open={open} onClose={handleClose}>
      <Title>Create a new record</Title>
      <Form onSubmit={formik.handleSubmit}>
        <TextField
          fullWidth
          name="amount"
          type="number"
          size="small"
          label="Amount"
          value={formik.values.amount}
          onBlur={formik.handleBlur}
          onChange={formik.handleChange}
          error={formik.touched.amount && !!formik.errors.amount}
          helperText={formik.touched.amount && formik.errors.amount}
        />
        <TextField
          select
          fullWidth
          name="type"
          size="small"
          label="Type"
          value={formik.values.type}
          onBlur={formik.handleBlur}
          onChange={formik.handleChange}
          error={formik.touched.type && !!formik.errors.type}
          helperText={formik.touched.type && formik.errors.type}
        >
          <MenuItem value="income">Income</MenuItem>
          <MenuItem value="expense">Expense</MenuItem>
        </TextField>
        <TextField
          select
          fullWidth
          name="category"
          size="small"
          label="Category"
          value={formik.values.category}
          onBlur={formik.handleBlur}
          onChange={formik.handleChange}
          error={formik.touched.category && !!formik.errors.category}
          helperText={formik.touched.category && formik.errors.category}
        >
          {categories.map((category) => (
            <MenuItem key={category.id} value={category.id}>
              {category.title}
            </MenuItem>
          ))}
        </TextField>
        <ButtonGroup>
          <Button
            sx={{ color: "#fff" }}
            variant="contained"
            color="success"
            type="submit"
          >
            Add record
          </Button>
          <Button
            sx={{ color: "#fff" }}
            variant="contained"
            color="error"
            onClick={handleClose}
          >
            Close
          </Button>
        </ButtonGroup>
      </Form>
    </ModalComponent>
  );
};
