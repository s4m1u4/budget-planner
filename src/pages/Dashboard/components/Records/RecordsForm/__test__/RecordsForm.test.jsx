import { RecordsForm } from "../RecordsForm";
import { act, render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { CATEGORIES } from "./__mock__";

const renderComponent = () => {
  return render(
    <RecordsForm
      open={open}
      onSubmit={onSubmit}
      categories={CATEGORIES}
      handleClose={handleClose}
      setNewHistory={setNewHistory}
    />
  );
};

const renderComponentAsync = async () => {
  await act(
    async () =>
      await render(
        <RecordsForm
          open={open}
          onSubmit={onSubmit}
          categories={CATEGORIES}
          handleClose={handleClose}
          setNewHistory={setNewHistory}
        />
      )
  );
};

const open = true;
const onSubmit = jest.fn();
const handleClose = jest.fn();
const setNewHistory = jest.fn();

describe("RecordsForm component", () => {
  it("snapshot matching", () => {
    const { baseElement } = renderComponent();
    expect(baseElement).toMatchSnapshot();
  });

  describe("should render", () => {
    beforeEach(() => {
      renderComponent();
    });

    it("the title", () => {
      expect(
        screen.getByRole("heading", { name: /create a new record/i })
      ).toBeInTheDocument();
    });

    it('an input with name="amount"', () => {
      expect(screen.getByLabelText(/amount/i)).toBeInTheDocument();
    });

    it('an input with name="type"', () => {
      expect(screen.getByLabelText(/type/i)).toBeInTheDocument();
    });

    it('an input with name="category"', () => {
      expect(screen.getByLabelText(/category/i)).toBeInTheDocument();
    });

    it('a button "Add record"', () => {
      expect(
        screen.getByRole("button", { name: /add record/i })
      ).toBeInTheDocument();
    });

    it('a button "Close"', () => {
      expect(
        screen.getByRole("button", { name: /close/i })
      ).toBeInTheDocument();
    });

    it("error message when submitting a form with empty inputs", async () => {
      expect(screen.queryByText(/amount is required/i)).toBeNull();
      expect(screen.queryByText(/type is required/i)).toBeNull();
      expect(screen.queryByText(/category is required/i)).toBeNull();

      userEvent.click(screen.getByRole("button", { name: /add record/i }));

      await renderComponentAsync();

      expect(screen.getByText(/amount is required/i)).toBeInTheDocument();
      expect(screen.getByText(/type is required/i)).toBeInTheDocument();
      expect(screen.getByText(/category is required/i)).toBeInTheDocument();
    });
  });

  it("successful form submission", async () => {
    await renderComponentAsync();

    userEvent.type(screen.getByLabelText(/amount/i), "1756");
    userEvent.click(screen.getByLabelText(/type/i));
    userEvent.click(screen.getByRole("option", { name: /income/i }));
    userEvent.click(screen.getByLabelText(/category/i));
    userEvent.click(screen.getByRole("option", { name: /work/i }));

    userEvent.click(screen.getByRole("button", { name: /add record/i }));

    await waitFor(() =>
      expect(onSubmit).toHaveBeenCalledWith({
        amount: 1756,
        type: "income",
        category: "084a3558-6b14-4e38-9d24-47e912021384",
      })
    );

    expect(onSubmit).toHaveBeenCalledTimes(1);
  });
});
