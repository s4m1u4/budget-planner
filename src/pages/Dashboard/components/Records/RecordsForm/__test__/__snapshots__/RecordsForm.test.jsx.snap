// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecordsForm component snapshot matching 1`] = `
<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="MuiModal-root css-79ws1d-MuiModal-root"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root css-i9fmh8-MuiBackdrop-root-MuiModal-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-test="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiBox-root css-5fyqun"
      tabindex="-1"
    >
      <h3
        class="css-ngqp4j"
      >
        Create a new record
      </h3>
      <form
        class="css-ezycoh"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-1pysi21-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="mui-1"
            id="mui-1-label"
          >
            Amount
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
              id="mui-1"
              name="amount"
              type="number"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-173wfxe"
              >
                <span>
                  Amount
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-1pysi21-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="mui-2"
            id="mui-2-label"
          >
            Type
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
            variant="outlined"
          >
            <div
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-labelledby="mui-2-label mui-2"
              class="MuiSelect-select MuiSelect-outlined MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-jedpe8-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
              id="mui-2"
              role="button"
              tabindex="0"
            >
              <span
                class="notranslate"
              >
                ​
              </span>
            </div>
            <input
              aria-hidden="true"
              class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
              name="type"
              tabindex="-1"
              value=""
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
              data-testid="ArrowDropDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-173wfxe"
              >
                <span>
                  Type
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-1pysi21-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="mui-3"
            id="mui-3-label"
          >
            Category
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
            variant="outlined"
          >
            <div
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-labelledby="mui-3-label mui-3"
              class="MuiSelect-select MuiSelect-outlined MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-jedpe8-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
              id="mui-3"
              role="button"
              tabindex="0"
            >
              <span
                class="notranslate"
              >
                ​
              </span>
            </div>
            <input
              aria-hidden="true"
              class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
              name="category"
              tabindex="-1"
              value=""
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
              data-testid="ArrowDropDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-173wfxe"
              >
                <span>
                  Category
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="css-20ow0d"
        >
          <button
            class="MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-wqakyb-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="submit"
          >
            Add record
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            class="MuiButton-root MuiButton-contained MuiButton-containedError MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-p8fncj-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Close
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </form>
    </div>
    <div
      data-test="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
